### PR TITLE
LAB-230-implement-settings-struct

### PR DESCRIPTION
**Pull request by YoniEisipsSNN**
Temporarily pointing to Lab-229 as base for ease of review.

### CODE REVIEW

File: src/impedance_analyzer/types.jl [hunk 1]
```diff
@@ -5,13 +5,22 @@ Supported Instruments:
...
+    settings::ImpedanceAnalyzerSettings
     connected::Bool
 end
-create_instrument(model::Type{<:ImpedanceAnalyzerModel}, address) = ImpedanceAnalyzer{model}(address, TCPSocket(), false)
 
+function create_instrument(model::Type{<:ImpedanceAnalyzerModel}, address)
+    ImpedanceAnalyzer{model}(
+        TCPInfo(address, TCPSocket()),
+        ImpedanceAnalyzerSettings(),
+        false,
```
Comment (mofii) [inline src/impedance_analyzer/types.jl:21]:
```julia
 connected = false
 ImpedanceAnalyzer{model}(
        TCPInfo(address, TCPSocket()),
        ImpedanceAnalyzerSettings(),
        connected,
 )
```

File: src/impedance_analyzer/types.jl [hunk 2]
```diff
@@ -5,13 +5,22 @@ Supported Instruments:
   - [`Agilent4395A`](@ref)
 """
 abstract type ImpedanceAnalyzerModel <: AbstractModel end
+mutable struct ImpedanceAnalyzerSettings <: AbstractSettings
```
Comment (mofii) [inline src/impedance_analyzer/types.jl:8]: I'd like to make the settings struct immutable in order to make it harder for the users to manually change settings

File: src/power_supply/types.jl [hunk 3]
```diff
@@ -7,12 +7,22 @@ Supported Instruments:
   - [`Versatile10010XRLXI`](@ref)
 """
 abstract type PowerSupplyModel <: AbstractModel end
+mutable struct PowerSupplySettings <: AbstractSettings
```
Comment (mofii) [inline src/power_supply/types.jl:10]: immutable.
What are we gaining from having an AbstractSettings type?

File: src/source_measure_unit/types.jl [hunk 4]
```diff
@@ -3,13 +3,23 @@ Supported Instruments:
 
   - [`AgilentB2910BL`](@ref)
 """
-abstract type SMUModel <: AbstractModel end
-mutable struct SourceMeasureUnit{SMUModel} <: AbstractInstrument
-    address::String
-    sock::TCPSocket
+abstract type SourceMeasureUnitModel <: AbstractModel end
+mutable struct SourceMeasureUnitSettings <: AbstractSettings
```
Comment (mofii) [inline src/source_measure_unit/types.jl:7]: immutable

File: src/waveform_generator/types.jl [hunk 5]
```diff
@@ -4,12 +4,22 @@ Supported Instruments:
   - [`Keysight33612A`](@ref)
 """
 abstract type WaveformGeneratorModel <: AbstractModel end
+mutable struct WaveformGeneratorSettings <: AbstractSettings
```
Comment (mofii) [inline src/waveform_generator/types.jl:7]: immutable

File: src/instrument.jl [hunk 6]
```diff
@@ -21,24 +21,36 @@ Use the help feature for documentation of each of the instrument groups for more
 """
 abstract type AbstractInstrument end
 abstract type AbstractModel end
+abstract type AbstractSettings end
+
 # This can be renamed later after some discussion :)
 mutable struct TCPInfo
```
Comment (mofii) [inline src/instrument.jl:27]: immutable

Comment (mofii) [reply src/instrument.jl:27]: The "connected" field should also go into this struct.

File: src/instrument.jl [hunk 7]
```diff
@@ -21,24 +21,36 @@ Use the help feature for documentation of each of the instrument groups for more
...
+end
+DefaultSettings() = DefaultSettings("")
+
 get_address(i::AbstractInstrument) = i.info.address
 get_socket(i::AbstractInstrument) = i.info.socket
 set_address(i::AbstractInstrument, a::String) = (i.info.address = a)
 set_socket(i::AbstractInstrument, s::TCPSocket) = (i.info.socket = s)
 
 mutable struct Instrument{AbstractModel} <: AbstractInstrument
     info::TCPInfo
```
Comment (mofii) [inline src/instrument.jl:43]: I think "info" is too vague a field name.
"tcp" could work.
Then it reads as"
```julia
instr.tcp.connected
instr.tcp.socket
instr.tcp.address
```

File: src/instrument.jl [hunk 8]
```diff
@@ -21,24 +21,36 @@ Use the help feature for documentation of each of the instrument groups for more
...
     address::String
     socket::TCPSocket
 end
 
+mutable struct DefaultSettings <: AbstractSettings
+    name::String
+end
+DefaultSettings() = DefaultSettings("")
+
 get_address(i::AbstractInstrument) = i.info.address
```
Comment (mofii) [inline src/instrument.jl:37]: "i" is often a counting index variable. 
"instr" is not super elegant but I like it better as an abbreviation for instrument than "i". Also open for other proposals!

**Review by mofii — CHANGES_REQUESTED** [review]
I'd like the settings structs to be immutable in order to make it harder for the users to manually change settings which then don't reflect the state of the instruments. 
The immutable structs can then internally be updated using the Accessors.jl package.
This could look something like this (untested):

```julia
function set_address!(instr::AbstractInstrument, address::String) 
 tcp = instr.tcp
 new_tcp = @set tcp.address = address
 instr.tcp = new_tcp
 return nothing
end
```

File: src/power_supply/types.jl (hunk 3 above)
Comment (gita135) [reply src/power_supply/types.jl]: @mofii  This is something I am unsure of, I can run it by Yoni once he returns 👍🏽

**Comment by mofii** [comment]
@gita135 
Let's keep commits that are related to the TCPInfo struct in the LAB-229 branch and PR. From a brief overview, it looks like that would be the following commits:
- Added set_tcp_connected!() calls to connect!() and close!()
- Implemented set_tcp_connected!()
- Fixed connected var in create_instrument()
- Added connected field to TCPInfo struct
- Added connected var to all uses of TcpInfo in create_instrument()
- Added connected var call to TcpInfo for instr
- Added Accessors Package
- Added connected var to create_instrument() call
- Refactored set_socket() and set_address() to use Accessors
- Added ! to set_address() and set_socket()

Please cherry pick the commits such they become part of LAB-229 and drop the commits here using `git rebase -i`.

I think the Accossors.jl code gets to be too clunky for now. Let's make the structs mutable and keep the code clean. We might at a later time be able to make a cleaner implementation of immutable structs Accessors.jl if needed.

**Comment by mofii** [comment]
Hi @YoniEisipsSNN and @gita135, 
This one has a lot of merge conflicts. Would you mind rebasing it and solving the merge conflicts?
Thanks!

**Review by mofii — CHANGES_REQUESTED** [review]
Needs a rebase

**Review by mofii — APPROVED** [review]
🥳

**Pull request merged**